transport/grpc/integration_test.go

```diff
@@ -299,6 +301,226 @@ func TestCustomContextDial(t *testing.T) {
+			wantMetrics: []metric{
+				{82, tagsCompression},
+				{82, tagsDecompression},
+				{23, tagsCompression},
+				{23, tagsDecompression},
+				{29, tagsCompression},
+				{29, tagsDecompression},
+				{75, tagsCompression},
+				{75, tagsDecompression},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.msg, func(t *testing.T) {
+			_metrics.reset()
+
+			tt.testEnvOptions.DialOptions = []DialOption{Compressor(tt.compressor)}
+			tt.do(t, func(t *testing.T, e *testEnv) {
+				value := strings.Repeat("a", 32*1024)
+				err := e.SetValueYARPC(context.Background(), "foo", value)
+				if tt.wantErr != "" {
+					assert.Error(t, err)
+					assert.EqualError(t, err, tt.wantErr)
+				} else if assert.NoError(t, err) {
+					getValue, err := e.GetValueYARPC(context.Background(), "foo")
+					assert.NoError(t, err)
+					assert.Equal(t, value, getValue)
+				}
+			})
+
+			compressor := ""
+			if tt.compressor != nil {
+				compressor = tt.compressor.Name()
+			}
+			assert.Equal(t, newMetrics(tt.wantMetrics, map[string]string{
+				"compressor": compressor,
+			}), _metrics)
+		})
+	}
+}
+
+type metricCollection struct {
+	metrics []metric
+}
+
+func (c *metricCollection) reset() {
+	c.metrics = c.metrics[:0]
+}
+
+func newMetrics(metrics []metric, tags map[string]string) *metricCollection {
+	c := metricCollection{
+		metrics: make([]metric, len(metrics)),
+	}
+	for i, m := range metrics {
+		c.metrics[i] = metric{
+			bytes: m.bytes,
+			tags:  map[string]string{},
+		}
+		for key, value := range m.tags {
+			c.metrics[i].tags[key] = value
+		}
+		for key, value := range tags {
+			c.metrics[i].tags[key] = value
+		}
+	}
+	return &c
+}
+
+type metric struct {
+	bytes int
+	tags  map[string]string
+}
+
+func (m *metric) Increment(value int) {
+	m.bytes += value
+}
+
+// new creates a new metrics data point and passes returns it as one element slice
+func (c *metricCollection) new(stage, compressor string) *metric {
+	l := len(c.metrics)
+	c.metrics = append(c.metrics, metric{
+		bytes: 0,
+		tags: map[string]string{
+			"compressor": compressor,
+			"stage":      stage,
+		},
+	})
+	return &c.metrics[l]
+}
+
+type counter interface {
+	Increment(value int)
+}
+
+type testCompressor struct {
+	name       string
+	metrics    *metricCollection
+	comperr    error
+	decomperr  error
+	enableGZip bool
+}
+
+func newCompressor(name string, metrics *metricCollection) *testCompressor {
+	comp := testCompressor{
+		name:    name,
+		metrics: metrics,
+	}
+
+	switch strings.TrimPrefix(name, "test-") {
+	case "fail-comp":
+		comp.comperr = assert.AnError
+	case "fail-decomp":
+		comp.decomperr = assert.AnError
+	case "gzip":
+		comp.enableGZip = true
+	}
+	return &comp
+}
+
+func (c *testCompressor) Name() string { return c.name }
+
+func (c *testCompressor) Compress(w io.Writer) (io.WriteCloser, error) {
+	metered := byteMeter{
+		Writer:  w,
+		counter: c.metrics.new("compress", c.name),
+	}
+
+	if !c.enableGZip {
```

Comment:
optional nit: I think it'd be clearer to put the special case flag with the functionality. ie swapping the returns and if-condition
```
if c.enableGZip {
  return gzip.NewWriter(&metered), nil
}
return &metered, c.comperr
```